llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp

```diff
@@ -6176,6 +6176,57 @@ static SDValue PerformLOADCombine(SDNode *N,
       DL);
 }
 
+static SDValue
+PerformBUILD_VECTORCombine(SDNode *N, TargetLowering::DAGCombinerInfo &DCI) {
+  auto VT = N->getValueType(0);
+  if (!DCI.isAfterLegalizeDAG() || !Isv2x16VT(VT))
+    return SDValue();
+
+  auto Op0 = N->getOperand(0);
+  auto Op1 = N->getOperand(1);
+
+  // Start out by assuming we want to take the lower 2 bytes of each i32
+  // operand.
+  uint64_t Op0Bytes = 0x10;
+  uint64_t Op1Bytes = 0x54;
+
+  std::pair<SDValue *, uint64_t *> OpData[2] = {{&Op0, &Op0Bytes},
+                                                {&Op1, &Op1Bytes}};
+
+  // Check that each operand is an i16, truncated from an i32 operand. We'll
+  // select individual bytes from those original operands. Optionally, fold in a
+  // shift right of that original operand.
+  for (auto &[Op, OpBytes] : OpData) {
+    // Eat up any bitcast
+    if (Op->getOpcode() == ISD::BITCAST)
+      *Op = Op->getOperand(0);
+
+    if (Op->getValueType() != MVT::i16 || Op->getOpcode() != ISD::TRUNCATE ||
+        Op->getOperand(0).getValueType() != MVT::i32)
+      return SDValue();
+
+    *Op = Op->getOperand(0);
+
+    // Optionally, fold in a shift-right of the original operand and permute
+    // the two higher bytes from the shifted operand
+    if (Op->getOpcode() == ISD::SRL && isa<ConstantSDNode>(Op->getOperand(1))) {
+      if (cast<ConstantSDNode>(Op->getOperand(1))->getZExtValue() == 16) {
+        *OpBytes += 0x22;
+        *Op = Op->getOperand(0);
+      }
+    }
```

Comment:
This seems like a separate optimization to me: `prmt a (shr b [8|16|24]) x` -> `prmt a b y`.

Could we implement a separate DAG combine to do this? That way, `LowerBUILD_VECTOR` (and eventually `LowerINSERT_VECTOR_ELT` and `LowerEXTRACT_VECTOR_ELT` once they begin to use `prmt`) can use it.

Reply:
Piggybacking off of this, if we do perform `prmt a (shr b [8|16|24]) x` -> `prmt a b y` separately, then should this be a custom lowering instead of a DAG combine?

Reply:
> This seems like a separate optimization to me: prmt a (shr b [8|16|24]) x -> prmt a b y.

One way we could generalize that is via two-step process:
- convert shifts and truncates by 8/16/24, vector extraction of i8/i16 into `prmt`, This give us convenient access to an i32 and a mask of valid bits in it. E.g. `shr b,16` -> `prmt b, 0, 0x4410`.
- then combine individual i32s with partial values into one. E.g. `or(prmt(a,0,0x4410), prmt(b,0,0x1044), 16))` -> `prmt(a,b,0x5410)`. It would be applicable to `prmt` itself, `or`, `build_vector`, `insert_vector_elt`, etc.

Reply:
My original approach was custom-lower v2x16 BUILD_VECTORs to PRMT, but I found it was getting too complicated to identify the cases that were optimal to use with PRMT. As discussed in another comment thread, unless the original values are i32, this optimization broadly isn't worth it. During lowering of BUILD_VECTOR, we often don't as easily see the 16-bit-from-32-bit values that make it worthwhile. My concern with being too eager about BUILD_VECTOR, is that there will necessarily be cases where we need to undo a suboptimal PRMT, which feels wrong to me.

I imagine that Artem's idea could work, however. I suspect we'll still have to undo an eagerly matched PRMT, though. Perhaps we'd have to be more careful with one-use nodes?

Reply:
How do you see this idea working in the case of something like `generic_2xi16` [here](https://github.com/llvm/llvm-project/blob/main/llvm/test/CodeGen/NVPTX/load-store.ll#L197-L217)?

In this case, it's better to extract to actual 16-bit registers from the vector using `mov.b32 {%rs1, %rs2}, %r1`, because we're doing scalarized 16-bit operations on them. We can't detect that at the point at which we do `LowerEXTRACT_VECTOR_ELEMENT`. Well we can, but it's not common to have to look at uses when lowering nodes, and it's a bit of a red flag to me. The problem with eagerly creating PRMTs when lowering the vector extracts is that it gives us these awkward 32-bit registers, which we have to undo.

We'd probably have to do what `NVPTXDAGToDAGISel::tryEXTRACT_VECTOR_ELEMENT` is doing and try and find two `truncate`s from two `PRMT`s from the same original vector and replace them all with a single `NVPTX::I32ToV2I16`. This starts to sound a little strange, especially since we haven't proven that using PRMTs in this way is going to bring any tangible benefits.

What do you think?

Reply:
The '32-bit-source + mask' intermadiate value approach would be able to preserve that nice example. Nothing stops us from lowering a pair of `SM(value, 0xffff), SM(value, 0xffff0000)` as `mov.s32 {rs1, rs2}, value` when we need two 16-bit values, but use `prmt` when the destination is 32-bit. E.g. if your example above would only swap those 16-bit parts, without addition.

Reply:
I suppose so, yes.

I was also wondering whether we'd need a way of encapsulating the idea of an "unused" byte for maximum effect, as in the example of truncation: `i16 = trunc (i32 a)` -> `prmt a, b, 0xXX10`. There's no concrete byte of `a` or `b` to select for the truncated-out bytes, nor any concrete value of `b` - both can be anything we want. If we choose any concrete values or indices, it might hinder or confuse optimal combines/lowering. Using something like `undef` as `b` would work, and maybe selecting any byte from an `undef` operand would give us enough info to know we can replace them with anything.

In any case, I won't have the time to dedicate to trying out any of these ideas, I'm afraid. If you want to take it on, go ahead! We could either merge this PR as-is, or just abandon it. I'm easy.

Reply:
You can always use zero as one of the PRMT inputs, or sign-extend upper bits.

This particular change is in the nice-to-have category but it's not particularly important. It's more of a "how we could produce slightly better code with a cleaner/nicer implementation", but given that things already work reasonably well the upside is very limited. It may be a suitable starter project for someone getting up to speed with LLVM/NVPTX in the future.